### ISSUE TITLE
Diffenator2 review

### ISSUE THREAD

**Issue by MariannaPaszkowska** [googlefonts/googlesans-flex#1040]
Reviewed masters: 576

- 545 masters with no regressions, and one new character ( Combining Short Solidus Overlay glyph, ( ̷  U+0337))

- [x]  1 master with slightly moved Ậ in QUẬN
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-0_0_ROND-0_0_slnt--10_0-diffenator.html

- [x]  11 masters with changes in the dollar sign:
-- opsz-6_0_wght-518_98_wdth-25_0_GRAD-100_0_ROND-100_0_slnt-0_0
-- opsz-6_0_wght-520_01_wdth-25_0_GRAD-100_0_ROND-100_0_slnt-0_0
-- opsz-6_0_wght-699_01_wdth-25_0_GRAD-100_0_ROND-100_0_slnt-0_0
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-100_0_ROND-100_0_slnt-0_0
-- opsz-18_0_wght-1_0_wdth-25_0_GRAD-100_0_ROND-0_0_slnt--10_0
-- opsz-18_0_wght-1_0_wdth-25_0_GRAD-100_0_ROND-0_0_slnt-0_0
-- opsz-18_0_wght-1_0_wdth-25_0_GRAD-100_0_ROND-100_0_slnt--10_0
-- opsz-144_0_wght-1_0_wdth-25_0_GRAD-100_0_ROND-0_0_slnt--10_0
-- opsz-144_0_wght-1_0_wdth-25_0_GRAD-100_0_ROND-0_0_slnt-0_0
-- opsz-144_0_wght-1_0_wdth-25_0_GRAD-100_0_ROND-100_0_slnt--10_0
-- opsz-144_0_wght-1_0_wdth-25_0_GRAD-100_0_ROND-100_0_slnt-0_0

example:
<img width="959" alt="Screenshot 2024-09-13 at 3 28 57 PM" src="https://github.com/user-attachments/assets/711c1d7b-805b-4c88-91b8-714fc9fee674">
<img width="959" alt="Screenshot 2024-09-13 at 3 29 03 PM" src="https://github.com/user-attachments/assets/cdcf9a50-951f-4723-a545-8b5af78ca682"> 

- [x] 20 masters with reported changes but I cannot confirm them visually. Kerning pairs that changed EQ VQ PQ (1 unit difference):
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-0_0_ROND-0_0_slnt--10_0
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-0_0_ROND-0_0_slnt-0_0
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-0_0_ROND-100_0_slnt--10_0
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-0_0_ROND-100_0_slnt-0_0
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-100_0_ROND-0_0_slnt--10_0
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-100_0_ROND-0_0_slnt-0_0
-- opsz-6_0_wght-700_0_wdth-25_0_GRAD-100_0_ROND-100_0_slnt--10_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-0_0_ROND-0_0_slnt--10_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-0_0_ROND-0_0_slnt-0_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-0_0_ROND-100_0_slnt--10_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-0_0_ROND-100_0_slnt-0_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-100_0_ROND-0_0_slnt--10_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-100_0_ROND-0_0_slnt-0_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-100_0_ROND-100_0_slnt--10_0
-- opsz-144_0_wght-699_01_wdth-100_0_GRAD-100_0_ROND-100_0_slnt-0_0
-- opsz-144_0_wght-700_0_wdth-100_0_GRAD-0_0_ROND-0_0_slnt--10_0
-- opsz-144_0_wght-700_0_wdth-100_0_GRAD-0_0_ROND-100_0_slnt-0_0
-- opsz-144_0_wght-700_0_wdth-100_0_GRAD-100_0_ROND-0_0_slnt-0_0
-- opsz-144_0_wght-700_0_wdth-100_0_GRAD-100_0_ROND-100_0_slnt--10_0
-- opsz-144_0_wght-700_0_wdth-100_0_GRAD-100_0_ROND-100_0_slnt-0_0

example
<img width="945" alt="Screenshot 2024-09-13 at 3 36 15 PM" src="https://github.com/user-attachments/assets/b89e26c7-9d80-4386-8053-29792197845b">

**Comment by MariannaPaszkowska** [comment]
@octaviopardo could you have a look at those three tasks and let me know if this was intentional? I don't recall planning the changes in dollar and Ậ ?

**Comment by octaviopardo** [comment]
Apologies. The changes in the dollar are indeed intentional. We detected an issue with the thickness of the glyph and the bar, we changed it and did not include it in the log to report it in this version (to be honest I thought we had done it for 2.003 already, not in the hiatus in between last and this next release). It will not happen again. Visual details of the dollar changes, top new version, bottom 2.003 version

![Captura de pantalla 2024-09-13 a las 18 14 02](https://github.com/user-attachments/assets/2b1d0fda-3643-4183-9e16-f8ecd9660b43)
![Captura de pantalla 2024-09-13 a las 18 13 53](https://github.com/user-attachments/assets/5794a301-7773-48b7-a114-45e44b072e77)


![Captura de pantalla 2024-09-13 a las 18 16 47](https://github.com/user-attachments/assets/1152c5af-165a-43f4-a56b-17a5797c031b)
![Captura de pantalla 2024-09-13 a las 18 16 38](https://github.com/user-attachments/assets/1f95f938-2438-497d-9691-059c4a909d3b)
![Captura de pantalla 2024-09-13 a las 18 16 21](https://github.com/user-attachments/assets/e201b0f4-96f2-4748-9b86-0eff432d94f1)
![Captura de pantalla 2024-09-13 a las 18 16 11](https://github.com/user-attachments/assets/2f496444-31aa-414a-9fcd-0b320ca3a463)
![Captura de pantalla 2024-09-13 a las 18 15 41](https://github.com/user-attachments/assets/751a53d3-d769-42cb-b21d-0782f98bc51e)
![Captura de pantalla 2024-09-13 a las 18 15 08](https://github.com/user-attachments/assets/79557559-a9e2-460a-9adb-612cdaa28b20)
![Captura de pantalla 2024-09-13 a las 18 14 51](https://github.com/user-attachments/assets/f6ccef7c-299a-4f24-a8f4-5b1966e1f313)
![Captura de pantalla 2024-09-13 a las 18 14 12](https://github.com/user-attachments/assets/df88ac38-a543-4078-ac9f-369740b83ff8)

**Comment by octaviopardo** [comment]
Regarding kerning
We reviewed the kerning of both sources (2.003 & 2.004). Kerning is identical in both versions in the sources.

Moreover! if you notice the coordinates where you are reporting the changes, they do not correspond with any master! (masters are always placed at weight400 and weight 1000; those coordinates might match some of the brace layers of the dollar/cent/eth.

To confirm that the issue was not in the sources I actually reviewed manually the kerning in the adjacent masters to the reported coordinates and there are no changes

In conclusion, I think the issue must be somewhere else

**Comment by Hoolean** [comment]
Very strange...

@MariannaPaszkowska Do you have a link to the workflow that generated the test fonts and the report? I want to see what the raw kerning is in the TTF to narrow it down further 🔎

**Comment by MariannaPaszkowska** [comment]
@Hoolean it was the font from this import: https://github.com/googlefonts/googlesans-flex/actions/runs/10795123116 

I run Diffenator2 locally, but I also now run it here: https://github.com/googlefonts/googlesans-flex/actions/runs/10898255616 (v2.003 vs import-GSF-Ashler-v2.4)

**Comment by Hoolean** [comment]
Thanks Marianna; I'll look into the kerning now. In the meantime, I have made a test PR that includes the changes from the import branch above as UFOs, but with the outlines for /dollar reverted to those in v2.003. The CI jobs are running now at #1043 if you would like to examine it :)

**Comment by Hoolean** [comment]
I can confirm that these are the only files touched in the UFO sources of the current import:

- GoogleSansFlex.designspace
- dollar.glif
- fontinfo.plist
- lib.plist
- order.plist
- periodcentered.loclC_A_T_.case.glif
- periodcentered.loclC_A_T_.glif
- sterling.tf.glif

This rules out a kerning change from the import process, or a minor outline difference in one of the involved glyphs causing different rounding.

I suspect we will find that the cause is either:
1. A change in the kerning in the final TTF, because of some difference we haven't pinned down yet; or
2. A change in the way that the browser rasterises /Q, now that it is made up of components, even if it has the same point structure.

I will start digging into the TTF to narrow down further.

**Comment by Hoolean** [comment]
In the TTF, the difference comes from an advance width difference in 'Q' at this location:

```
$ hb-shape.exe Before/GoogleSansFlex\[GRAD\,ROND\,opsz\,slnt\,wdth\,wght\].ttf 'Q' --variations="opsz=144,wght=700,wdth=100,GRAD=0,ROND=100,slnt=0"
[Q=0+1577]

$ hb-shape.exe After/GoogleSansFlex\[GRAD\,ROND\,opsz\,slnt\,wdth\,wght\].ttf 'Q' --variations="opsz=144,wght=700,wdth=100,GRAD=0,ROND=100,slnt=0"
[Q=0+1578]
```

**Comment by Hoolean** [comment]
In v2.003, /Q has an advance width stored at the position of the /dollar and /cent sparse layer, even though neither it nor its components are in the layer themselves in our sources:

```
$ python interesting_hvar_locs.py Before/GoogleSansFlex\[GRAD\,ROND\,opsz\,slnt\,wdth\,wght\].ttf Q
HVAR tuples:
opsz [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0), (0.0, 1.0, 1.0)]
wdth [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0), (0.0, 1.0, 1.0)]
wght [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0), (0.0, 0.19830322265625, 1.0), (0.19830322265625, 0.20001220703125, 1.0), (0.20001220703125, 1.0, 1.0)]
GRAD [(0.0, 0.0, 0.0)]
ROND [(0.0, 0.0, 0.0)]
slnt [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0)]

HVAR locations:
opsz [-1.0, 0.0, 1.0]
wdth [-1.0, 0.0, 1.0]
wght [-1.0, 0.0, 0.19830322265625, 0.20001220703125, 1.0]
GRAD [0.0]
ROND [0.0]
slnt [-1.0, 0.0]
```

In the latest sources, after re-composing /Q, these advance widths are gone:

```
$ python interesting_hvar_locs.py After/GoogleSansFlex\[GRAD\,ROND\,opsz\,slnt\,wdth\,wght\].ttf Q
HVAR tuples:
opsz [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0), (0.0, 1.0, 1.0)]
wdth [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0), (0.0, 1.0, 1.0)]
wght [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0), (0.0, 1.0, 1.0)]
GRAD [(0.0, 0.0, 0.0)]
ROND [(0.0, 0.0, 0.0)]
slnt [(-1.0, -1.0, 0.0), (0.0, 0.0, 0.0)]

HVAR locations:
opsz [-1.0, 0.0, 1.0]
wdth [-1.0, 0.0, 1.0]
wght [-1.0, 0.0, 1.0]
GRAD [0.0]
ROND [0.0]
slnt [-1.0, 0.0]
```

(tested with the test script now pushed to #1044)

**Comment by Hoolean** [comment]
The new behaviour is more correct, as it matches what we have in the sources.

I suspect that fonttools or ufo2ft is over-eager when automatically decomposing glyphs, and creates metrics (or even outlines) for them at every sparse layer position, even those where they do not contribute. As coordinates must be rounded wherever they are defined, this introduces minor differences: /Q would not otherwise have rounded coordinates at the position of /cent and /dollar's sparse layer, and so this introduces the risk of the shaping engine rounding differently.

**Comment by Hoolean** [comment]
We just used the script in #1044 to compare every glyph in v2.003 against the new font, and only /Q has different `HVAR` designspace positions before and after.

We were able to identify that /Q's `HVAR` off-by-one was just a side-effect of an outline off-by-one, however, so we would need to adapt the script to inspect `gvar` instead to assert that no other glyphs were affected.

**Comment by Hoolean** [comment]
I've updated the script, and can now confirm that only /Q is affected by the issue we identified; no other glyphs in the TTFs have gained or lost outlines at the sparse dollar/cent layer positions in the new release 🎉 If we decompose /Q again, then it should eliminate the issue we identified entirely.

**Comment by Hoolean** [comment]
Finally tracked down the root cause of this! 🤯

### Details

`_tail.Q` has brace layers at wght=520 but `O` does not. When we were decomposing `Q`, ufo2ft had to instantiate **rounded** `O` outlines at wght=520 to make the decomposed `Q` at that position.

Now that we are not decomposing `Q`, the glyphs `_tail.Q` and `O` can be used as components as-is, and so ufo2ft no longer has to instantiate the `O` part of the `Q` at this position. This means it is **unrounded**, and may be rounded by the shaper differently.

This was not a compiler bug; it was just necessary to round the outlines at that position when we needed to put a decomposed glyph there.

Additionally, the `dollar` and `cent` brace layer was a red herring; this just happened to be where diffenator2 spotted the rounding difference. No extra outlines for `Q` glyphs are being created at this position.

### Summary

With this and the analysis above, we can confirm that `Q` was the only glyph affected, and that decomposing it again will resolve the pixel diff with the previous release. There is also no compiler bug to report, as ufo2ft was doing the right thing.

**Comment by MariannaPaszkowska** [comment]
@Hoolean I ran diffenator2 with higher precision and threshold and I got many more letters reported as modified. I need to double-check, but it seems it now reports the glyphs that are rebuilt with components and also some others. Those are glyphs that have changed pixels 0.1 - 0.0001% (--threshold 0.0001 --precision 56). So far, I've reviewed 170 locations reported by diffenator2 out of 484.
`$`, U+0024
`0`, U+0030
`Q`, U+0051
`Ð`, U+00D0
`÷`, U+00F7
`Đ`, U+0110
`Ħ`, U+0126
`Ŀ`, U+013F
`ŀ`, U+0140
`Ł`, U+0141
`ł`, U+0142
`Ơ`, U+01A0
`ơ`, U+01A1
`Ư`, U+01AF
`Ẩ`, U+1EA8
`Ẳ`, U+1EB2
`Ể`, U+1EC2
`Ổ`, U+1ED4
`Ớ`, U+1EDA
`ớ`, U+1EDB
`Ờ`, U+1EDC
`ờ`, U+1EDD
`Ở`, U+1EDE
`ở`, U+1EDF
`Ỡ`, U+1EE0
`ỡ`, U+1EE1
`Ợ`, U+1EE2
`ợ`, U+1EE3
`Ứ`, U+1EE8
`Ừ`, U+1EEA
`Ử`, U+1EEC
`Ữ`, U+1EEE
`Ự`, U+1EF0
`≠`, U+2260
`≤`, U+2264
`ﬀ`, U+FB00
`ﬁ`, U+FB01
`ﬃ`, U+FB03
`ﬄ`, U+FB04

**Comment by Hoolean** [comment]
Ah no... :(

I think these will be caused by a different issue to /Q; what locations are the glyphs broken at? (just one or two is enough to poke around)

At this precision it could be that composite glyphs are unavoidably rendered differently, but it would be good for us to work out exactly why this is happening and rule out any outline issues first.

**Comment by MariannaPaszkowska** [comment]
no regressions found: 

diffenator2: https://github.com/googlefonts/googlesans-flex/actions/runs/11317784987
testing workspace fonts: https://github.com/googlefonts/googlesans-flex/actions/runs/11281708901
